### DISCUSSION TITLE
Where should CSS live?

### DISCUSSION THREAD

**Discussion by dannyvalle** [nuxt/nuxt#17761] [General]
As I am learning Nuxt, specifically V3, I am confused as to the right way to add styles. Usually I use SASS and add a link to include my complied CSS file. However, with Nuxt, you can add styles globally or scoped within VUE files in the "style" block. 

- Is this the preferred way to do it?
- What's the advantage, disadvantage?
- Should I use a combination of both?

**Comment by Aareksio** [comment]
As always, it depends on the project you are building. With Nuxt, you are almost certain to use Single File Components. As such you should use `<style>` block to contain your components' styles:

```html
<template>
  <p class="my-paragraph">Hello World</p>
</template>

<style lang="scss" scoped>
  .my-paragraph {
    color: red;
  }
</style>
```

You may notice the use of `lang` and `scoped` attributes:
- `lang` allows you to specify the pre-processor for the block
- `scoped` limits the scope of your style so it does not bleed outside the component

As for **global** styles, you may want to include them in separate `.scss` file and include in `nuxt.config.ts` `css` field:

```ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  css: [
    '@/assets/styles/index.scss'
  ]
})
```

You may want to define variables, mixins and other _resources_ for global use. If using Vite, I believe this is the current way to do this:

```ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
  vite: {
    css: {
      preprocessorOptions: {
        scss: {
          additionalData: '@import \'@/assets/styles/resources.scss\';'
        }
      }
    },
  }
})
```

This adds an import to EACH of your component files. You should **not** include any code which compiles to CSS in `resources.scss`. Only pre-processor macros such as variables, functions and mixins may be used. Otherwise you'll end up bundling a copy of CSS with each component.

---

Why component styles:
- include only the styles you actually need on the page, styles for unused components are not downloaded
- style lives next to the component, it is easy to track HTML it belongs to
- IDE can help with finding unused styles, it helps to navigate between HTML and CSS too
- you can use `scoped` styles, to avoid CSS bleeding - it'll add component id to your CSS selectors, so they do not match anything outside
- maintenance - you can delete components along with the styles when no longer needed
- DX - the scope of style is limited to single component, no need to search for global classes

When to use global styles:
- native components - paragraphs, headers, lists
- anything that can not be a component - html, body, fonts, [role], ::selection, scrollbar
- outside HTML - 3rd party plugins not nested inside components (eg. Chat Bubble)
- animation keyframes and transitions

Alternative approach - utility framework, such as tailwind. Do not write styles at all. It works well with component driven architecture. Not my personal choice, but it has huge following.

**Reply by dannyvalle** [reply]
Wow, thank you for the detailed explanation. This really made things click. Thank you!
